File: scripts/yapc2012/yoshizow.js
// -*- javascript -*-

var primeSum = function(nth) {
    if (nth <= 5) throw new RangeError();

    var primes = new Array(nth);
    primes.splice(0, 5,  2, 3, 5, 7, 11);
    var count = 5;
    var sum = 28;
outer:
    for (var p = primes[count - 1] + 2; ; p += 2) {
        for (var i = 0; i < count; i++) {
            var q = primes[i];
            if (q * q > p)
                break;
            if (p % q == 0)
                continue outer;
        }
        primes[count] = p;
        count++;
        sum += p;
        if (count >= nth)
            break;
    }

    return sum;
}

var result = primeSum(10000);
console.log(result);

File: scripts/yapc2012/yoshizow.rb
# -*- ruby -*-

include Math

def prime_sum(nth)
  raise 'range error' if nth <= 5

  sieve_size = (nth * log(nth) + nth * log(log(nth)) / 2).floor
  sieve = Array.new(sieve_size)  # sieve[i] is for 2*i+1
  count = 1
  sum = 2
  p = nil
  q = 1
  while true
    while sieve[q]
      q += 1
    end
    p = 2 * q + 1
    # p p
    count += 1
    sum += p
    if count >= nth
      return sum
    end
    i = q
    while i < sieve_size
      sieve[i] = true
      i += p
    end
  end
end

#nth = if ARGV.length > 0 then ARGV[0].to_i else 10000 end
nth = 10000
result = prime_sum(nth)
p result

